### PR TITLE
enhancement(wasm transform): Implement some UX improvements for WASM

### PR DESCRIPTION
**Pull request by Hoverbear**
While working on https://github.com/timberio/vector-website/pull/67, as a plugin author, I wanted a couple features.

Notably:

* A structured WASM config, instead of just a JSON blob.
* Corrects an issue where `max_heap_memory_size` was not configurable.
* More comments in the first example I will try to build a plugin based on.

### CODE REVIEW

**Comment by Hoverbear** [comment]
Ah! I had one more thing to add, I wanted to make the demo `add_fields` take options.

**Comment by Hoverbear** [comment]
@JeanMertz If you're back, do you think you could review this? :)

File: tests/data/wasm/add_fields/src/lib.rs [hunk 1]
```diff
@@ -1,31 +1,87 @@
...
     1
 }
 
+/// Perform one-time optional shutdown events.
+///
+/// **Note:** There is no guarantee this function will be called before shutdown,
+/// as we may be forcibly killed.
 #[no_mangle]
-pub extern "C" fn shutdown() {}
+pub extern "C" fn shutdown() {}
```
Comment (JeanMertz) [inline tests/data/wasm/add_fields/src/lib.rs:87]: nit: missing EOF newline

**Review by JeanMertz — APPROVED** [review]
nice improvements 👍

File: tests/data/wasm/add_fields/src/lib.rs [hunk 2]
```diff
@@ -1,31 +1,87 @@
...
 
+/// Process data starting from a given point in memory to another point.
+///
+/// It's not necessary for the plugin to actually read, or parse this data.
+///
+/// Call [`vector_wasm::hostcall::emit`] to emit a message out.
+///
+/// # Returns
+///
+/// This function should return the number of emitted messages.
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply tests/data/wasm/add_fields/src/lib.rs:43]: Hmm I think I should say it's a "hint"

**Comment by Hoverbear** [comment]
#3606 is affecting Mac here.

**Pull request merged**